geode-core/src/main/java/org/apache/geode/management/internal/beans/stats/GCStatsMonitor.java

```diff
@@ -39,69 +39,71 @@
   private volatile long collections = 0;
   private volatile long collectionTime = 0;
 
-  long getCollections() {
-    return collections;
-  }
-
-  long getCollectionTime() {
-    return collectionTime;
-  }
-
   public GCStatsMonitor(String name) {
     super(name);
   }
 
-  void decreasePrevValues(Map<String, Number> statsMap) {
-    collections -= statsMap.getOrDefault(StatsKey.VM_GC_STATS_COLLECTIONS, 0).longValue();
-    collectionTime -= statsMap.getOrDefault(StatsKey.VM_GC_STATS_COLLECTION_TIME, 0).longValue();
-  }
-
-  void increaseStats(String name, Number value) {
-    if (name.equals(StatsKey.VM_GC_STATS_COLLECTIONS)) {
-      collections += value.longValue();
-      return;
-    }
-
-    if (name.equals(StatsKey.VM_GC_STATS_COLLECTION_TIME)) {
-      collectionTime += value.longValue();
-      return;
+  @Override
+  // this will be called multiple times with different collector's stats
+  public void addStatisticsToMonitor(Statistics stats) {
+    monitor.addListener(this);// if already listener is added this will be a no-op
+    monitor.addStatistics(stats);
+
+    // stats map should keep the sum of all the GC stats
+    StatisticDescriptor[] descriptors = stats.getType().getStatistics();
+    for (StatisticDescriptor d : descriptors) {
+      String name = d.getName();
+      Number value = stats.get(d);
+      if (name.equals(StatsKey.VM_GC_STATS_COLLECTIONS)) {
+        collections += value.longValue();
+      } else if (name.equals(StatsKey.VM_GC_STATS_COLLECTION_TIME)) {
+        collectionTime += value.longValue();
+      }
     }
   }
 
   @Override
   public Number getStatistic(String statName) {
     if (statName.equals(StatsKey.VM_GC_STATS_COLLECTIONS)) {
-      return getCollections();
+      return collections;
     }
 
     if (statName.equals(StatsKey.VM_GC_STATS_COLLECTION_TIME)) {
-      return getCollectionTime();
+      return collectionTime;
     }
-
     return 0;
   }
 
   @Override
   public void handleNotification(StatisticsNotification notification) {
-    decreasePrevValues(statsMap);
-
+    // sum up all the count and all the time in the stats included in this notification
+    long totalCount = 0;
+    long totalTime = 0;
     for (StatisticId statId : notification) {
       StatisticDescriptor descriptor = statId.getStatisticDescriptor();
       String name = descriptor.getName();
       Number value;
-
       try {
         value = notification.getValue(statId);
       } catch (StatisticNotFoundException e) {
         value = 0;
       }
-
       log(name, value);
-      increaseStats(name, value);
-      statsMap.put(name, value);
+      if (name.equals(StatsKey.VM_GC_STATS_COLLECTIONS)) {
+        totalCount += value.longValue();
+      }
+
+      else if (name.equals(StatsKey.VM_GC_STATS_COLLECTION_TIME)) {
+        totalTime += value.longValue();
+      }
     }
+
+    collections = totalCount;
+    collectionTime = totalTime;
   }
```

Comment:
I think a notification includes only those stats whose values have changed. If that's correct, what we're summing here is only the _changed_ GC stats, not _all_ of the GC stats. This will work only if we're absolutely certain that every stat will change on every sample.

See `ValueMonitor.monitorStatistics()`, which builds the notification only from updated stats.

See `SampleCollector.sample()`, which adds a stats to the updated stats list only if the value of the stat changed from the previous sample.

Reply:
Please write unit tests for `GCSTatsMonitor` directly, rather than testing only through `MemberLevelStatsTest`.

Reply:
> I think a notification includes only those stats whose values have changed. If that's correct, what we're summing here is only the _changed_ GC stats, not _all_ of the GC stats. This will work only if we're absolutely certain that every stat will change on every sample.
> 
> See `ValueMonitor.monitorStatistics()`, which builds the notification only from updated stats.
> 
> See `SampleCollector.sample()`, which adds a stats to the updated stats list only if the value of the stat changed from the previous sample.

then what would those count/time mean though. It's not right to report just a single one of them. It's not right to sum them up either then....

Reply:
> Please write unit tests for `GCSTatsMonitor` directly, rather than testing only through `MemberLevelStatsTest`.

the unit test will be just like the MemberLevelStatsTest, only by mocking all those variables.

Reply:
> then what would those count/time mean though. It's not right to report just a single one of them. It's not right to sum them up either then....

The stats notification stuff is specifically designed to report only changes. As a result, old gen and new gen CGs are almost always reported independently (unless they both happen to change during a single sample).

I think that's why the old code was doing the weird stuff. It was trying to track the previous value for each kind of GC, so that when a new value arrived, it could add in just the change in that kind of GC. Decrementing and incrementing was a puzzling way to do that, but that was the goal.

Here's a way that seems simpler to me than decrementing/incrementing. I'll use collections as an example, but it's the same for collectionTime.
- Create a map that records the latest value for each GC type.
- Create a field that records the sum of values for all types. This is what gets reported via getCollections()
- When a new kind of GC is added via addStatisticsToMonitor()…
    1.  Store the initial value for that GC type in the map.
    2. Sum the values of all GC types in the map, and store the sum in the field.
- When a notification arrives…
    1. Store the new value for that GC type in the map.
    2. Sum the values of all GC types in the map, and store the sum in the field.

Reply:
Let me correct my terminology a bit… I keep saying "GC type," and that's not the right distinction. What distinguishes one batch of GC stats from another is which GarbageCollectorMXBean the values come from. I think (but I'm not sure) that each GC bean reports about a particular memory space.

So where I say "GC type," a better term might be "memory space".

Reply:
> > then what would those count/time mean though. It's not right to report just a single one of them. It's not right to sum them up either then....
> 
> The stats notification stuff is specifically designed to report only changes. As a result, old gen and new gen CGs are almost always reported independently (unless they both happen to change during a single sample).
> 
> I think that's why the old code was doing the weird stuff. It was trying to track the previous value for each kind of GC, so that when a new value arrived, it could add in just the change in that kind of GC. Decrementing and incrementing was a puzzling way to do that, but that was the goal.
> 
> Here's a way that seems simpler to me than decrementing/incrementing. I'll use collections as an example, but it's the same for collectionTime.
> 
> * Create a map that records the latest value for each GC type.
> * Create a field that records the sum of values for all types. This is what gets reported via getCollections()
> * When a new kind of GC is added via addStatisticsToMonitor()…
>   
>   1. Store the initial value for that GC type in the map.
>   2. Sum the values of all GC types in the map, and store the sum in the field.
> * When a notification arrives…
>   
>   1. Store the new value for that GC type in the map.
>   2. Sum the values of all GC types in the map, and store the sum in the field.

Yes, this sounds like a better idea. I will implement it.